File: server/src/lib/crockford-base32.ts
/* Crockford base32 (no padding) — encode-only. Used for the ephemeral pairing
   code and the 80-bit CA fingerprint tag in the companion pairing QR. Output is
   a subset of QR alphanumeric mode, so the QR stays in its densest encoding.
   Alphabet excludes I, L, O, U to avoid visual ambiguity in manual entry. */
const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

export function crockfordBase32(bytes: Uint8Array): string {
  let out = '';
  let buffer = 0;
  let bits = 0;
  for (const byte of bytes) {
    buffer = (buffer << 8) | byte;
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      out += ALPHABET[(buffer >>> bits) & 0x1f];
    }
  }
  if (bits > 0) {
    out += ALPHABET[(buffer << (5 - bits)) & 0x1f];
  }
  return out;
}

File: apps/android/lib/src/data/crockford_base32.dart
/// Crockford base32 (no padding), encode-only. Mirrors the server's
/// `crockford-base32.ts` so the companion can recompute the CA fingerprint tag
/// and compare it to the QR's `fpTag`. Alphabet excludes I, L, O, U.
const _alphabet = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

String crockfordBase32(List<int> bytes) {
  final out = StringBuffer();
  var buffer = 0, bits = 0;
  for (final byte in bytes) {
    buffer = (buffer << 8) | (byte & 0xff);
    bits += 8;
    while (bits >= 5) {
      bits -= 5;
      out.write(_alphabet[(buffer >> bits) & 0x1f]);
    }
  }
  if (bits > 0) out.write(_alphabet[(buffer << (5 - bits)) & 0x1f]);
  return out.toString();
}
